File: kotlin/day03.kt
package day03
import common.readInputLines

fun getMostCommon(nums: Iterable<List<Int>>, pos: Int): Int {
    var n0 = 0;
    var n1 = 0;
    for (num in nums) {
        n0 += 1 - num[pos];
        n1 += num[pos];
    }
    if (n1 >= n0) {
        return 1
    } else {
        return 0
    }
}

fun digitsToDecimal(digits: Iterable<Int>): Int {
    return digits.fold(0) {v, d -> v * 2 + d}
}


fun part1(nums: List<List<Int>>): Int {
    var lc = mutableListOf<Int>()
    var mc = mutableListOf<Int>()
    for (i in 0 until nums[0].size) {
        val m = getMostCommon(nums, i)
        mc.add(m)
        lc.add(1 - m)
    }
    return digitsToDecimal(mc) * digitsToDecimal(lc)
}

fun part2(nums: List<List<Int>>): Int {
    var mnums = (0 until nums.size).toList()
    var lnums = (0 until nums.size).toList()
    for (i in 0 until nums[0].size) {
        if (mnums.size > 1) {
            val m = getMostCommon(mnums.map {nums[it]}, i)
            mnums = mnums.filter {nums[it][i] == m }
        }
        if (lnums.size > 1) {
            val m = getMostCommon(lnums.map {nums[it]}, i)
            lnums = lnums.filter {nums[it][i] == 1 - m }
        }
    }
    return digitsToDecimal(nums[mnums[0]]) * digitsToDecimal(nums[lnums[0]])
}

fun solution() {
    val nums = readInputLines(3).map {it.toCharArray().map {Character.getNumericValue(it)}}
    println("Answer 1: ${part1(nums)}\nAnswer 2: ${part2(nums)}")
}

File: rust/src/day03.rs
use crate::common;

fn get_most_common<'a>(nums: impl Iterator<Item = &'a Vec<u8>>, pos: usize) -> u8 {
    let mut n0 = 0;
    let mut n1 = 0;
    for num in nums {
        n0 += (num[pos] == 0) as usize;
        n1 += (num[pos] == 1) as usize;
    }
    if n1 >= n0 {
        1
    } else {
        0
    }
}

fn part1(nums: &Vec<Vec<u8>>) -> i64 {
    let mut lc: i64 = 0;
    let mut mc: i64 = 0;
    for i in 0..nums[0].len() {
        let m = get_most_common(nums.iter(), i) as i64;
        mc = mc * 2 + m;
        lc = lc * 2 + (1 - m);
    }
    mc * lc
}

fn digits_to_decimal(digits: &Vec<u8>) -> i64 {
    digits.iter().fold(0i64, |v, d| v * 2 + *d as i64)
}

fn part2(nums: &Vec<Vec<u8>>) -> i64 {
    let mut mnums: Vec<usize> = (0..nums.len()).collect();
    let mut lnums: Vec<usize> = (0..nums.len()).collect();
    for i in 0..nums[0].len() {
        if mnums.len() > 1 {
            let m = get_most_common(mnums.iter().cloned().map(|x| &nums[x]), i);
            mnums = mnums.iter().cloned().filter(|&x| nums[x][i] == m).collect();
        }
        if lnums.len() > 1 {
            let l = 1 - get_most_common(lnums.iter().cloned().map(|x| &nums[x]), i);
            lnums = lnums.iter().cloned().filter(|&x| nums[x][i] == l).collect();
        }
    }
    digits_to_decimal(&nums[mnums[0]]) * digits_to_decimal(&nums[lnums[0]])
}

pub(crate) fn solution() {
    let nums: Vec<Vec<u8>> = common::read_lines(&common::data_file(3))
        .unwrap()
        .iter()
        .map(|line| {
            line.chars()
                .map(|x| x.to_digit(10).unwrap() as u8)
                .collect()
        })
        .collect();
    println!("Answer 1: {}\nAnswger 2: {}", part1(&nums), part2(&nums));
}
